File: mtiruman_HW2/Java/hw2.java
import java.io.*;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Created by raijin on 3/1/16.
 */
public class hw2 {

    public static void hw2(String inFile, String outFile) throws IOException{
        String line = null;
        try{
            FileReader fr = new FileReader(inFile);
            BufferedReader br = new BufferedReader(fr);
            FileWriter fw = new FileWriter(outFile);
            BufferedWriter bw = new BufferedWriter(fw);

            ArrayList stack = new ArrayList();
            while ((line = br.readLine()) != null){
            	   if (line.contains("quit"))
            	   	break;
            	   else
            	   	parse(line, stack);
            }
            for (int i = stack.size() - 1; i >= 0; i--){
                bw.write(stack.get(i) + "\n");
            }
            bw.close();
        }
        catch (FileNotFoundException e){
            e.printStackTrace();
        }
    }

    private static void parse(String command, ArrayList stack){
        if (command.contains(":error:")){
            //System.out.println(command);
            stack.add(":error:");
        }
        else if (command.contains(":true:")) {
            //System.out.println(command);
            stack.add(":true:");
        }
        else if (command.contains(":false:")) {
            //System.out.println(command);
            stack.add(":false:");
        }
        else if (command.contains("neg")) {
            //System.out.println(command);
            if (stack.size() == 0) {
                stack.add(":error:");
            }
            else if (stack.size() > 0){
                if (stack.get(stack.size() - 1) instanceof String) {
                    stack.add(":error:");
                }
                else if (stack.get(stack.size() - 1) instanceof Integer){
                    int a = (-1)*((int) stack.remove(stack.size() - 1));
                    stack.add(a);
                }
            }
        }
        else if (command.contains("push")){
            //String num = command.replaceAll("[^0-9?!\\.]","");
            String num = command.substring(5, command.length());
            if (num.contains("."))
                stack.add(":error:");
            else{
                int a = Integer.parseInt(num);
                stack.add(a);
            }
        }
        else if (command.contains("pop")){
            if (stack.size() == 0) {
                stack.add(":error:");
            }
            else if (stack.size() > 0){
                stack.remove(stack.size() - 1);
            }
        }
        else if (command.contains("add")){
            if (stack.size() <= 1) {
                stack.add(":error:");
            }
            else if (stack.size() > 1){
                int x, y;
                if ((stack.get(stack.size() - 1) instanceof Integer) && (stack.get(stack.size() - 2)) instanceof Integer) {
                    y = (int) stack.remove(stack.size() - 1);
                    x = (int) stack.remove(stack.size() - 1);
                    stack.add(y + x);
                }
                else{
                    stack.add(":error:");
                }
            }
        }
        else if (command.contains("sub")){
            if (stack.size() <= 1) {
                stack.add(":error:");
            }
            else if (stack.size() > 1){
                int x, y;
                if ((stack.get(stack.size() - 1) instanceof Integer) && (stack.get(stack.size() - 2)) instanceof Integer) {
                    y = (int) stack.remove(stack.size() - 1);
                    x = (int) stack.remove(stack.size() - 1);
                    stack.add(x - y);
                }
                else{
                    stack.add(":error:");
                }
            }
        }
        else if (command.contains("mul")){
            if (stack.size() <= 1) {
                stack.add(":error:");
            }
            else if (stack.size() > 1){
                int x, y;
                if ((stack.get(stack.size() - 1) instanceof Integer) && (stack.get(stack.size() - 2)) instanceof Integer) {
                    y = (int) stack.remove(stack.size() - 1);
                    x = (int) stack.remove(stack.size() - 1);
                    stack.add(y*x);
                }
                else{
                    stack.add(":error:");
                }
            }
        }
        else if (command.contains("div")){
            if (stack.size() <= 1) {
                stack.add(":error:");
            }
            else if (stack.size() > 1){
                int x, y;
                if ((stack.get(stack.size() - 1) instanceof Integer) && (stack.get(stack.size() - 2)) instanceof Integer) {
                    y = (int) stack.remove(stack.size() - 1);
                    x = (int) stack.remove(stack.size() - 1);
                    if (y == 0){
                        stack.add(x);
                        stack.add(y);
                        stack.add(":error:");
                    }
                    else {
                        stack.add(x/y);
                    }
                }
                else{
                    stack.add(":error:");
                }
            }
        }
        else if (command.contains("rem")){
            if (stack.size() <= 1) {
                stack.add(":error:");
            }
            else if (stack.size() > 1){
                int x, y;
                if ((stack.get(stack.size() - 1) instanceof Integer) && (stack.get(stack.size() - 2)) instanceof Integer) {
                    y = (int) stack.remove(stack.size() - 1);
                    x = (int) stack.remove(stack.size() - 1);
                    if (y == 0){
                        stack.add(x);
                        stack.add(y);
                        stack.add(":error:");
                    }
                    else {
                        stack.add(x%y);
                    }
                }
                else{
                    stack.add(":error:");
                }
            }
        }
        else if (command.contains("swap")){
            if (stack.size() <= 1) {
                stack.add(":error:");
            }
            else if (stack.size() > 1) {
                Object temp = stack.get(stack.size() - 1);
                stack.set(stack.size() - 1,stack.get(stack.size() - 2));
                stack.set(stack.size() - 2,temp);
            }
        }
    }
}

File: mtiruman_HW2/Python/hw2.py
import re

def hw2(inFile, outFile):
	inputFile = open(inFile, "r")
	outputFile = open(outFile, "w")
	stack = []
	for line in inputFile:
		if ":error:" in line:
			stack.append(":error:")
		elif ":true:" in line:
			stack.append(":true:")
		elif ":false:" in line:
			stack.append(":false:")
		elif "neg" in line:
			if len(stack) == 0:
				stack.append(":error:")
			elif len(stack) > 0:
				if isinstance(stack[-1], str):
					stack.append(":error:")
				elif isinstance(stack[-1], int):
					a = (-1)*(stack.pop())
					stack.append(a)
		elif "push" in line:
			num = re.findall("[-+]?\d+[\.]?\d*",line)
			if "." in num[0]:
				stack.append(":error:")
			else:
				stack.append(int(num[0]))
		elif "pop" in line:
			if len(stack) == 0:
				stack.append(":error:")
			else:
				stack.pop()
		elif "add" in line:
			if len(stack) <= 1:
				stack.append(":error:")
			else:
				if (isinstance(stack[-1], int) and isinstance(stack[-2], int)):
					y = int(stack.pop())
					x = int(stack.pop())
					stack.append(y+x)
				else:
					stack.append(":error:")
		elif "sub" in line:
			if len(stack) <= 1:
				stack.append(":error:")
			else:
				if (isinstance(stack[-1], int) and isinstance(stack[-2], int)):
					y = int(stack.pop())
					x = int(stack.pop())
					stack.append(x-y)
				else:
					stack.append(":error:")
		elif "mul" in line:
			if len(stack) <= 1:
				stack.append(":error:")
			else:
				if (isinstance(stack[-1], int) and isinstance(stack[-2], int)):
					y = int(stack.pop())
					x = int(stack.pop())
					stack.append(y*x)
				else:
					stack.append(":error:")
		elif "div" in line:
			if len(stack) <= 1:
				stack.append(":error:")
			else:
				if (isinstance(stack[-1], int) and isinstance(stack[-2], int)):
					y = int(stack.pop())
					x = int(stack.pop())
					if y == 0:
						stack.append(x)
						stack.append(y)
						stack.append(":error:")
					else:
						stack.append(x/y)
				else:
					stack.append(":error:")
		elif "rem" in line:
			if len(stack) <= 1:
				stack.append(":error:")
			else:
				if (isinstance(stack[-1], int) and isinstance(stack[-2], int)):
					y = int(stack.pop())
					x = int(stack.pop())
					if y == 0:
						stack.append(x)
						stack.append(y)
						stack.append(":error:")
					else:
						stack.append(x%y)
				else:
					stack.append(":error:")
		elif "swap" in line:
			if len(stack) <= 1:
				stack.append(":error:")
			else:
				temp = stack[-1]
				stack[-1] = stack[-2]
				stack[-2] = temp
		elif "quit" in line:
			break
	stack.reverse()
	for i in stack:
		outputFile.write(str(i) + "\n")